dynamodb/db.go

```diff
@@ -0,0 +1,226 @@
+	// "DynamoDB typically deletes expired items within 48 hours of expiration. The exact duration within
+	// which an item truly gets deleted after expiration is specific to the nature of the workload
+	// and the size of the table."
+	Expiration time.Time `dynamodbav:"expiration,unixtime"`
+	// Version is an internal field used to control flushing/draining the Value field concurrently
+	Version uint `dynamodbav:"version"`
+}
+
+func newDDBBucket(name string, expiration time.Time) ddbBucket {
+	return ddbBucket{
+		ddbBucketStatePrimaryKey: ddbBucketStatePrimaryKey{
+			Name: name,
+		},
+		Value:      0,
+		Expiration: expiration,
+		Version:    0,
+	}
+}
+
+func decodeBucket(b map[string]*dynamodb.AttributeValue) (*ddbBucket, error) {
+	var bs ddbBucket
+	if err := dynamodbattribute.UnmarshalMap(b, &bs); err != nil {
+		return nil, err
+	}
+	return &bs, nil
+}
+
+func encodeBucket(b ddbBucket) (map[string]*dynamodb.AttributeValue, error) {
+	return dynamodbattribute.MarshalMap(b)
+
+}
+
+func (b *ddbBucket) expired() bool {
+	return time.Now().After(b.Expiration)
+}
+
+func (db *bucketDB) key(name string) (map[string]*dynamodb.AttributeValue, error) {
+	return dynamodbattribute.MarshalMap(ddbBucketStatePrimaryKey{
+		Name: string(name),
+	})
+}
+
+func (db *bucketDB) bucket(name string) (*ddbBucket, error) {
+	key, err := db.key(name)
+	if err != nil {
+		return nil, err
+	}
+	res, err := db.ddb.GetItem(&dynamodb.GetItemInput{
+		Key:            key,
+		TableName:      aws.String(db.tableName),
+		ConsistentRead: aws.Bool(true),
+	})
+	if awsErr(err, dynamodb.ErrCodeResourceNotFoundException) || len(res.Item) == 0 {
+		return nil, errBucketNotFound
+	} else if err != nil {
+		return nil, err
+	}
+
+	return decodeBucket(res.Item)
+}
+
+func (db *bucketDB) createOrFindBucket(bucket ddbBucket) (*ddbBucket, error) {
+	data, err := encodeBucket(bucket)
+	if err != nil {
+		return nil, err
+	}
+	_, err = db.ddb.PutItem(&dynamodb.PutItemInput{
+		TableName: aws.String(db.tableName),
+		Item:      data,
+		ExpressionAttributeNames: map[string]*string{
+			"#N": aws.String("name"),
+		},
+		ConditionExpression: aws.String("attribute_not_exists(#N)"),
+	})
+	if err != nil {
+		if !awsErr(err, dynamodb.ErrCodeConditionalCheckFailedException) {
+			return nil, err
+		}
+		// for existing buckets simply fetch
+		return db.bucket(bucket.Name)
+	}
+
+	return &bucket, err
+}
+
+func (db *bucketDB) incrementBucketValue(name string, amount, capacity uint) (*ddbBucket, error) {
+	key, err := db.key(name)
+	if err != nil {
+		return nil, err
+	}
+	// https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.UpdateExpressions.html#Expressions.UpdateExpressions.SET.IncrementAndDecrement
+	res, err := db.ddb.UpdateItem(&dynamodb.UpdateItemInput{
+		Key:       key,
+		TableName: aws.String(db.tableName),
+		ExpressionAttributeValues: map[string]*dynamodb.AttributeValue{
+			":a": {
+				N: aws.String(strconv.FormatUint(uint64(amount), 10)),
```

Comment:
I had to look up this function and how it works, maybe consider something more familiar, e.g.
```suggestion
				N: aws.String(fmt.Sprintf("%d", amount)),
```

Reply:
AH! That is a great point. I'll go ahead with that change.

Reply:
https://github.com/Clever/leakybucket/pull/34/commits/725c7d8c8e0bb61d0a7c19acd3f393534ccde96f